### PR TITLE
 Add Fine tuning guide with KerasHub Gemma4 Model

### PR DESCRIPTION
**Pull request by laxmareddyp**

### CODE REVIEW

File: guides/keras_hub/text_classification_with_gemma4.py [hunk 1]
```diff
@@ -0,0 +1,709 @@
...
+catastrophic forgetting of the pretrained knowledge.
+
+**When to use full fine-tuning?** When you have sufficient memory and
+want maximum accuracy. The tradeoff is higher training cost and memory
+usage. Note that `gemma4_26b_a4b` has 26B total parameters (MoE), so
+full fine-tuning requires substantial resources.
+"""
+
+# Unfreeze the backbone for full fine-tuning.
+backbone.trainable = True
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply guides/keras_hub/text_classification_with_gemma4.py:324]: I will add the smallest preset as soon as it becomes available on KerasHub in Kaggle.

File: guides/keras_hub/text_classification_with_gemma4.py [hunk 2]
```diff
@@ -0,0 +1,709 @@
...
+    # Pad token_ids and padding_mask to MAX_SEQUENCE_LENGTH.
+    pad_length = MAX_SEQUENCE_LENGTH - real_length
+    token_ids = tf.pad(token_ids, [[0, pad_length]], constant_values=0)
+    padding_mask = tf.pad(padding_mask, [[0, pad_length]], constant_values=0)
+
+    # Ensure static shapes for batching.
+    token_ids = tf.ensure_shape(token_ids, [MAX_SEQUENCE_LENGTH])
+    padding_mask = tf.ensure_shape(padding_mask, [MAX_SEQUENCE_LENGTH])
+
+    return {"token_ids": token_ids, "padding_mask": padding_mask}, label
```
[the remark this answers is not in the document]
Comment (laxmareddyp) [reply guides/keras_hub/text_classification_with_gemma4.py:157]: Yes, the manual truncation/padding logic was unnecessarily verbose. One small correction on the suggested code . Sequence_length is a constructor parameter on SentencePieceTokenizer, not a call-time argument. Passing it as tokenizer(text, sequence_length=...) won't work, it needs to be set when the tokenizer is created.

The tokenizer then uses this internally during tokenize() to convert ragged output to a dense tensor with the specified length.

File: guides/keras_hub/text_classification_with_gemma4.py [hunk 3]
```diff
@@ -0,0 +1,689 @@
...
+    print("=" * 55)
+    print(f"{'Method':<30} {'Accuracy':>10}")
+    print("-" * 55)
+    for method, accuracy in results.items():
+        print(f"{method:<30} {accuracy:>10.4f}")
+    print("=" * 55 + "\n")
+
+
+def build_classifier(backbone, trainable_backbone=True):
+    """Build a binary classifier on top of the Gemma 4 backbone."""
```
Comment (divyashreepathihalli) [inline guides/keras_hub/text_classification_with_gemma4.py:203]: why are you building a new model here? we should show it with Gemma 4

File: guides/keras_hub/text_classification_with_gemma4.py [hunk 4]
```diff
@@ -0,0 +1,689 @@
...
+
+    model = keras.Model(
+        inputs={"token_ids": token_ids, "padding_mask": padding_mask},
+        outputs=output,
+    )
+    return model
+
+
+def predict_sentiment(model, text):
+    """Predict sentiment for a single text input."""
```
Comment (divyashreepathihalli) [inline guides/keras_hub/text_classification_with_gemma4.py:240]: we dont need separate function to show predict and inference. it should be part of the guide just explain - add code etc.

**Review by divyashreepathihalli — CHANGES_REQUESTED** [review]

**Comment by laxmareddyp** [comment]
@divyashreepathihalli Will re-write this complete guide Gemma4 capabilities along with fine tuning techniques in one guide.

File: scripts/hub_master.py [hunk 5]
```diff
@@ -3510,6 +3510,10 @@
         {
             "path": "segment_anything_in_keras_hub",
             "title": "Segment Anything",
+        },
+                {
```
Comment (hertschuh) [inline scripts/hub_master.py:3514]: Please fix indentation

File: guides/keras_hub/text_classification_with_gemma4.py [hunk 6]
```diff
@@ -0,0 +1,689 @@
...
+
+"""shell
+pip install -q --upgrade keras keras-hub tensorflow-datasets
+"""
+
+import os
+from google.colab import userdata
+
+os.environ["KERAS_BACKEND"] = "tensorflow"
+
```
Comment (hertschuh) [inline guides/keras_hub/text_classification_with_gemma4.py:85]: Just curious, why tensorflow and not JAX?

File: guides/keras_hub/text_classification_with_gemma4.py [hunk 7]
```diff
@@ -0,0 +1,689 @@
...
+- **Faster training**: Fewer gradients to compute.
+- **No inference overhead**: LoRA weights can be merged back into the
+  backbone after training.
+
+KerasHub makes enabling LoRA trivially easy with a one-line API.
+"""
+
+# Clean up previous models to free memory.
+del feature_extraction_model
+del full_finetune_model
```
Comment (hertschuh) [inline guides/keras_hub/text_classification_with_gemma4.py:371]: You probably need to also `del backbone` otherwise you're not freeing much memory.

File: guides/keras_hub/text_classification_with_gemma4.py [hunk 8]
```diff
@@ -0,0 +1,689 @@
...
+)
+
+print("\n--- Post-Training Quantization Evaluation ---")
+print_results_so_far()
+ptq_results = ptq_model.evaluate(test_ds)
+results["Post-Training Quantization"] = ptq_results[1]
+print(
+    f"Post-Training Quantization - Loss: {ptq_results[0]:.4f}, "
+    f"Accuracy: {ptq_results[1]:.4f}"
+)
```
Comment (hertschuh) [inline guides/keras_hub/text_classification_with_gemma4.py:484]: But you didn't train your classifier head, it's just random weights, so the results will be garbage.

Did you mean to reuse some classifier head you trained earlier?

File: guides/keras_hub/text_classification_with_gemma4.py [hunk 9]
```diff
@@ -0,0 +1,689 @@
...
+
+lora_q_model.compile(
+    optimizer=keras.optimizers.Adam(learning_rate=1e-4),
+    loss="binary_crossentropy",
+    metrics=["accuracy"],
+)
+
+print("\n--- LoRA Fine-Tuning + Quantization ---")
+print_results_so_far()
+lora_q_model.fit(
```
Comment (hertschuh) [inline guides/keras_hub/text_classification_with_gemma4.py:517]: Just curious why you so `print_results_so_far` before `model.fit`.

Wouldn't it be better after so that they're right next to the new result?
